### PR TITLE
[BEAM-14347] Add generic registration for accumulators

### PR DESCRIPTION
**Pull request by damccorm**
This is a continuation of the work started in #17429 and #17574 and extends the existing registration logic to allow registration of combiners.

Design doc: https://docs.google.com/document/d/1MLZLUhb-Sn7AJ666yn3TnYgMKE6wiXIBs7cjB_20hMA/edit?usp=sharing

After this, the big remaining chunks of work are:

1. Perf testing
2. Updates to the docs/our examples

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @lostluck 

Since this one has a bunch of generic + generated code and is a tougher one to review without context, I'm going to send this directly to you

File: sdks/go/pkg/beam/registration/registration.tmpl [hunk 1]
```diff
@@ -241,13 +242,320 @@ type teardown1x1 interface {
...
+type extractOutput1x1[T1, T2 any] interface {
+    ExtractOutput(a T1) T2
+}
+
+type extractOutput1x2[T1, T2 any] interface {
+    ExtractOutput(a T1) (T2, error)
+}
+
+{{range $accum := upto 3}}{{$genericParams := (add $accum 1)}}
+// Accumulator{{$genericParams}} registers an accumulator (CombineFn) DoFn's structural functions
```
Comment (lostluck) [inline sdks/go/pkg/beam/registration/registration.tmpl:278]: We probably want to call this Combiner1 etc.
They are called CombineFns or Combiners not accumulators. Accumulators are the intermediate type that CombineFns operate on. See https://beam.apache.org/documentation/programming-guide/#combine for the usual user nomenclature.

Eg. for a mean Combinefn, one could have (some number, say `int`) as AddInput's input type. AddInput would add it to the accumulator type `struct{Count int, Sum float64}`, MergeAccumulators merges them together, and ExtractOutput does the division of the Sum by the Count, to get the `float64` output type for the Mean.

File: sdks/go/pkg/beam/registration/registration.tmpl [hunk 2]
```diff
@@ -241,13 +242,320 @@ type teardown1x1 interface {
...
+		reflectx.RegisterFunc(reflect.TypeOf((*func(T0, T0) T0)(nil)).Elem(), caller)
+
+        mergeAccumulatorsWrapper = func(fn interface{}) reflectx.Func {
+			return reflectx.MakeFunc(func(a0 T0, a1 T0) T0 {
+				return fn.(mergeAccumulators2x1[T0]).MergeAccumulators(a0, a1)
+			})
+		}
+    }
+
+    if mergeAccumulatorsWrapper == nil{
```
Comment (lostluck) [inline sdks/go/pkg/beam/registration/registration.tmpl:320]: need space after nil

**Review by lostluck — CHANGES_REQUESTED** [review]
Naming aside: LGTM.

File: sdks/go/pkg/beam/registration/registration.tmpl (hunk 2 above)
Comment (damccorm) [reply sdks/go/pkg/beam/registration/registration.tmpl]: Done!

File: sdks/go/pkg/beam/registration/registration.tmpl (hunk 1 above)
Comment (damccorm) [reply sdks/go/pkg/beam/registration/registration.tmpl]: Yeah, that makes sense - updated!

**Review by lostluck — APPROVED** [review]
Will merge once the conflict from the emitters/iters is resolved.

**Comment by damccorm** [comment]
> Will merge once the conflict from the emitters/iters is resolved.

Done

**Pull request merged**